src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1107,7 +1103,11 @@ static void uct_dc_mlx5_iface_handle_failure(uct_ib_iface_t *ib_iface,
                                  UCS_MASK(UCT_IB_QPN_ORDER);
     uint8_t             dci    = uct_dc_mlx5_iface_dci_find(iface, qp_num);
 
-    iface->tx.dcis[dci].ops->handle_failure(iface, cqe, dci, status);
+    if (!uct_dc_mlx5_iface_is_dci_keepalive(iface, dci)) {
+        uct_dc_mlx5_dci_handle_failure(iface, cqe, dci, status);
+    } else {
+        uct_dc_mlx5_dci_keepalive_handle_failure(iface, cqe, dci, status);
+    }
```

Comment:
maybe invert order to avoid ! (we prefer to simplify this if this is not fast path)